// @switchboard-xyz/common/lib/CrossbarClient.cjs
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossbarClient = void 0;
const instructions_js_1 = require("./utils/instructions.cjs");
const string_js_1 = require("./utils/string.cjs");
const axios_1 = __importDefault(require("axios"));
const bs58_1 = __importDefault(require("bs58"));
class CrossbarClient {
    static default(verbose) {
        if (!CrossbarClient._instance) {
            CrossbarClient._instance = new CrossbarClient('https://crossbar.switchboard.xyz', verbose);
        }
        return CrossbarClient._instance;
    }
    constructor(crossbarUrl, verbose) {
        // feed hash -> crossbar response
        this.feedCache = new Map();
        this.crossbarUrl = new URL(crossbarUrl).origin;
        this.verbose = !!verbose;
    }
    /**
     * GET /fetch/:feedHash
     * Fetch data from the crossbar using the provided feedHash
     * @param {string} feedHash - The hash of the feed to fetch data for
     * @returns {Promise<{feedHash: string; queueHex: string; jobs: IOracleJob[];}} - The data fetched from the crossbar
     */
    async fetch(feedHash) {
        try {
            // Check if the feedHash is already in the cache
            const cached = this.feedCache.get(feedHash);
            if (cached)
                return cached;
            // Fetch the data from the crossbar
            console.log(`${this.crossbarUrl}/fetch/${feedHash}`);
            const response = await axios_1.default
                .get(`${this.crossbarUrl}/fetch/${feedHash}`)
                .then(resp => resp.data);
            // Cache the response on the crossbar instance
            this.feedCache.set(feedHash, response);
            return response;
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            // If response is outside of the 200 range, log the status and throw an error.
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetch status: ${response.status}`);
        }
    }
    /**
     * POST /store
     * Store oracle jobs on the crossbar, associated with a queue address
     * @param {string} queueAddress - The address of the queue
     * @param {IOracleJob[]} jobs - The oracle jobs to store
     * @returns {Promise<{ cid: string; feedHash: string; queueHex: string }>} - The stored data information
     */
    async store(queueAddress, jobs) {
        try {
            // Try to decode the queueAddress to a Buffer so that we can send it in the expected format,
            // base58, to the Crossbar node.
            const queue = (0, string_js_1.decodeString)(queueAddress);
            if (!queue)
                throw new Error(`Unable to parse queue: ${queueAddress}`);
            return await axios_1.default
                .post(`${this.crossbarUrl}/store`, { queue: bs58_1.default.encode(queue), jobs }, { headers: { 'Content-Type': 'application/json' } })
                .then(resp => {
                if (resp.status === 200)
                    return resp.data;
                throw new Error(`Bad Crossbar store response: ${resp.status}`);
            });
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar store response: ${response.status}`);
        }
    }
    /**
     * GET /simulate/:feedHashes
     * Simulate fetching feed results from the crossbar using feed hashes
     * @param {string[]} feedHashes - The hashes of the feeds to simulate
     * @param {boolean} [includeReceipts] - Whether to include receipts in the response
     * @returns {Promise<CrossbarSimulateResponse[]>} - The simulated feed results
     */
    async simulateFeeds(feedHashes, includeReceipts) {
        try {
            if (!feedHashes || feedHashes.length === 0)
                throw new Error('At least one feed is required');
            const feedsParam = feedHashes.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/simulate/${feedsParam}`, {
                params: { includeReceipts },
            })
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar simulateFeed response: ${response.status}`);
        }
    }
    /**
     * GET /updates/solana/:network/:feedpubkeys
     * Fetch updates for Solana network feeds from the crossbar
     * @param {string} network - The Solana network to fetch updates for
     * @param {string[]} feedpubkeys - The public keys of the feeds to fetch updates for
     * @param {number} [numSignatures] - The number of signatures to fetch (optional)
     * @returns {Promise<{ success: boolean; pullIx: TransactionInstruction; responses: { oracle: string; result: number | null; errors: string }[]; lookupTables: string[] }[]>} - The updates for the specified feeds
     */
    async fetchSolanaUpdates(network, feedpubkeys, payer, numSignatures) {
        try {
            if (!network)
                throw new Error('Network is required');
            if (!feedpubkeys || feedpubkeys.length === 0)
                throw new Error('At least one feed is required');
            const feedsParam = feedpubkeys.join(',');
            const response = await axios_1.default
                .get(`${this.crossbarUrl}/updates/solana/${network}/${feedsParam}`, {
                params: { numSignatures, payer },
            })
                .then(resp => resp.data);
            // Convert pullIx from hex to TransactionInstruction using IxFromHex
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const updates = response.map((update) => ({
                ...update,
                pullIxns: update.pullIxns.map(instructions_js_1.IxFromHex),
            }));
            return updates;
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchSolanaUpdates response: ${response.status}`);
        }
    }
    /**
     * GET /simulate/solana/:network/:feedpubkeys
     * Simulate fetching Solana feed results from the crossbar
     * @param {string} network - The Solana network to simulate
     * @param {string[]} feedpubkeys - The public keys of the feeds to simulate
     * @returns {Promise<{ feed: string; feedHash: string; results: number[] }[]>} - The simulated feed results
     */
    async simulateSolanaFeeds(network, feedpubkeys) {
        try {
            if (!network)
                throw new Error('Network is required');
            if (!feedpubkeys || feedpubkeys.length === 0)
                throw new Error('At least one feed is required');
            const feedsParam = feedpubkeys.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/simulate/solana/${network}/${feedsParam}`)
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar simulateSolanaFeeds response: ${response.status}`);
        }
    }
    /**
     * GET /updates/evm/:chainId/:aggregatorIds
     * Fetch updates for EVM network feeds from the crossbar
     * @param param0 - The chain ID and aggregator IDs to fetch updates for
     * @returns Promise<{ results: EVMResult[]; encoded: string[] }> - The updates for the specified feeds
     */
    async fetchEVMResults({ chainId, aggregatorIds, }) {
        try {
            if (!chainId)
                throw new Error('Chain ID is required');
            if (!aggregatorIds || aggregatorIds.length === 0)
                throw new Error('At least one feed is required');
            const feedsParam = aggregatorIds.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/updates/evm/${chainId}/${feedsParam}`)
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchEVMUpdates response: ${response.status}`);
        }
    }
    /**
     * GET /simulate/evm/:network/:aggregatorIds
     * Simulate fetching Solana feed results from the crossbar
     * @param {string} network - The Solana network to simulate
     * @param {string[]} aggregatorIds - The public keys of the feeds to simulate
     * @returns {Promise<{ feed: string; feedHash: string; results: number[] }[]>} - The simulated feed results
     */
    async simulateEVMFeeds(network, aggregatorIds) {
        try {
            if (!network)
                throw new Error('Network is required');
            if (!aggregatorIds || aggregatorIds.length === 0)
                throw new Error('At least one feed is required');
            const feedsParam = aggregatorIds.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/simulate/evm/${network}/${feedsParam}`)
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar simulateEVMFeeds response: ${response.status}`);
        }
    }
    /**
     * GET /randomness/evm/:chainId/:randomnessId
     * @param param0 - The chain ID and randomness ID to resolve
     */
    async resolveEVMRandomness({ chainId, randomnessId, }) {
        try {
            return await axios_1.default
                .get(`${this.crossbarUrl}/randomness/evm/${chainId}/${randomnessId}`)
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar resolveEVMRandomness response: ${response.status}`);
        }
    }
    /**
     * GET /updates/sui/:network/:aggregatorAddresses
     * Fetch updates for Sui network feeds from the crossbar
     * @param {string} network - The Sui network to fetch updates for (mainnet / testnet)
     * @param {string[]} aggregatorAddresses - The addresses of the aggregators to fetch updates for
     * @returns {Promise<{ responses: SuiAggregatorResponse[]; failures: string[] }>} - The updates and any failures for the specified feeds
     */
    async fetchSuiUpdates(network, aggregatorAddresses) {
        try {
            if (!network)
                throw new Error('Network is required');
            if (!aggregatorAddresses || aggregatorAddresses.length === 0)
                throw new Error('At least one aggregator address is required');
            const addressesParam = aggregatorAddresses.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/updates/sui/${network}/${addressesParam}`)
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchSuiUpdates response: ${response.status}`);
        }
    }
    /**
     * GET /simulate/sui/:network/:aggregatorAddresses
     * Simulate fetching Sui feed results from the crossbar
     * @param {string} network - The Sui network to simulate (mainnet / testnet)
     * @param {string[]} aggregatorAddresses - The feed IDs to simulate
     * @param {boolean} [includeReceipts] - Whether to include receipts in the response
     * @returns {Promise<SuiSimulationResult[]>} - The simulated feed results
     */
    async simulateSuiFeeds(network, aggregatorAddresses, includeReceipts) {
        try {
            if (!network)
                throw new Error('Network is required');
            if (!aggregatorAddresses || aggregatorAddresses.length === 0)
                throw new Error('At least one feed ID is required');
            const feedsParam = aggregatorAddresses.join(',');
            return await axios_1.default
                .get(`${this.crossbarUrl}/simulate/sui/${network}/${feedsParam}`, {
                params: { includeReceipts },
            })
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar simulateSuiFeeds response: ${response.status}`);
        }
    }
    /**
     * GET /gateways
     * Fetch all gateways from the crossbar
     * @param {string} [network] - Optional network parameter (devnet/testnet/mainnet). Defaults to mainnet.
     * @returns {Promise<string[]>} - The gateways response containing an array of Gateway urls
     */
    async fetchGateways(network = 'mainnet') {
        try {
            console.log(`${this.crossbarUrl}/gateways?network=${network}`);
            const gateways = await axios_1.default
                .get(`${this.crossbarUrl}/gateways?network=${network}`)
                .then(resp => resp.data);
            return gateways;
        }
        catch (err) {
            console.error('Error fetching gateways:', err);
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchGateways response: ${response.status}`);
        }
    }
    /**
     * GET /oracles
     * Fetch all oracles for a given network
     * @param {string} [network] - Optional network parameter (devnet/mainnet). Defaults to mainnet.
     * @returns {Promise<Oracle[]>} - Array of oracle information
     */
    async fetchOracles(network) {
        try {
            console.log(`${this.crossbarUrl}/oracles?network=${network}`);
            return await axios_1.default
                .get(`${this.crossbarUrl}/oracles`, {
                params: { network },
            })
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchOracles response: ${response.status}`);
        }
    }
    /**
     * POST /oracles/fetch_signatures
     * Fetch signatures from oracles for a given set of jobs
     * @param {FetchSignaturesRequest} request - The request parameters
     * @param {string} [network] - Optional network parameter (devnet/mainnet). Defaults to mainnet.
     * @returns {Promise<FeedEvalResponse[]>} - Array of oracle signatures and results
     */
    async fetchSignatures(request, network) {
        try {
            return await axios_1.default
                .post(`${this.crossbarUrl}/oracles/fetch_signatures`, request, {
                params: { network },
            })
                .then(resp => resp.data);
        }
        catch (err) {
            if (!axios_1.default.isAxiosError(err))
                throw err;
            const response = err.response;
            if (!response)
                throw err;
            if (this.verbose)
                console.error(`${response.status}: ${response.data}`);
            throw new Error(`Bad Crossbar fetchOracleSignatures response: ${response.status}`);
        }
    }
}
exports.CrossbarClient = CrossbarClient;
CrossbarClient._instance = null;
