// @blockworks-foundation/mangolana/lib/types.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockHeightStrategyClass = exports.TimeStrategyClass = exports.SequenceType = exports._TransactionInstructionWithIndex = exports.TransactionInstructionWithSigners = void 0;
var TransactionInstructionWithSigners = /** @class */ (function () {
    function TransactionInstructionWithSigners(transactionInstruction, signers) {
        if (signers === void 0) { signers = []; }
        this.transactionInstruction = transactionInstruction;
        this.signers = signers;
    }
    return TransactionInstructionWithSigners;
}());
exports.TransactionInstructionWithSigners = TransactionInstructionWithSigners;
var _TransactionInstructionWithIndex = /** @class */ (function (_super) {
    __extends(_TransactionInstructionWithIndex, _super);
    function _TransactionInstructionWithIndex(transactionInstruction, signers, index) {
        if (signers === void 0) { signers = []; }
        var _this = _super.call(this, transactionInstruction, signers) || this;
        _this.index = index;
        return _this;
    }
    return _TransactionInstructionWithIndex;
}(TransactionInstructionWithSigners));
exports._TransactionInstructionWithIndex = _TransactionInstructionWithIndex;
var SequenceType;
(function (SequenceType) {
    SequenceType[SequenceType["Sequential"] = 0] = "Sequential";
    SequenceType[SequenceType["Parallel"] = 1] = "Parallel";
    SequenceType[SequenceType["StopOnFailure"] = 2] = "StopOnFailure";
})(SequenceType = exports.SequenceType || (exports.SequenceType = {}));
var TimeStrategyClass = /** @class */ (function () {
    function TimeStrategyClass(_a) {
        var _b = _a.timeout, timeout = _b === void 0 ? 90 : _b, _c = _a.getSignatureStatusesPoolIntervalMs, getSignatureStatusesPoolIntervalMs = _c === void 0 ? 2000 : _c;
        this.timeout = timeout;
        this.getSignatureStatusesPoolIntervalMs = getSignatureStatusesPoolIntervalMs;
    }
    return TimeStrategyClass;
}());
exports.TimeStrategyClass = TimeStrategyClass;
var BlockHeightStrategyClass = /** @class */ (function () {
    function BlockHeightStrategyClass(_a) {
        var _b = _a.startBlockCheckAfterSecs, startBlockCheckAfterSecs = _b === void 0 ? 90 : _b, block = _a.block, _c = _a.getSignatureStatusesPoolIntervalMs, getSignatureStatusesPoolIntervalMs = _c === void 0 ? 2000 : _c;
        this.startBlockCheckAfterSecs = startBlockCheckAfterSecs;
        this.block = block;
        this.getSignatureStatusesPoolIntervalMs = getSignatureStatusesPoolIntervalMs;
    }
    return BlockHeightStrategyClass;
}());
exports.BlockHeightStrategyClass = BlockHeightStrategyClass;
