// @blockworks-foundation/mango-v4/dist/cjs/src/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SBOD_ORACLE_LUTS = exports.DEFAULT_RECEIVER_PROGRAM_ID = exports.JUPITER = exports.MAX_RECENT_PRIORITY_FEE_ACCOUNTS = exports.USDC_MINT = exports.MANGO_V4_MAIN_GROUP = exports.MANGO_V4_ID = exports.OPENBOOK_PROGRAM_ID = exports.COMPUTE_BUDGET_PROGRAM_ID = exports.RUST_I64_MIN = exports.RUST_I64_MAX = exports.RUST_U64_MAX = void 0;
const anchor_1 = require("@coral-xyz/anchor");
const web3_js_1 = require("@solana/web3.js");
const RUST_U64_MAX = () => {
    return new anchor_1.BN('18446744073709551615');
};
exports.RUST_U64_MAX = RUST_U64_MAX;
const RUST_I64_MAX = () => {
    return new anchor_1.BN('9223372036854775807');
};
exports.RUST_I64_MAX = RUST_I64_MAX;
const RUST_I64_MIN = () => {
    return new anchor_1.BN('-9223372036854775807');
};
exports.RUST_I64_MIN = RUST_I64_MIN;
exports.COMPUTE_BUDGET_PROGRAM_ID = new web3_js_1.PublicKey('ComputeBudget111111111111111111111111111111');
exports.OPENBOOK_PROGRAM_ID = {
    devnet: new web3_js_1.PublicKey('EoTcMgcDRTJVZDMZWBoU6rhYHZfkNTVEAfz3uUJRcYGj'),
    'mainnet-beta': new web3_js_1.PublicKey('srmqPvymJeFKQ4zGQed1GFppgkRHL9kaELCbyksJtPX'),
};
exports.MANGO_V4_ID = {
    testnet: new web3_js_1.PublicKey('4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg'),
    devnet: new web3_js_1.PublicKey('4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg'),
    'mainnet-beta': new web3_js_1.PublicKey('4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg'),
};
exports.MANGO_V4_MAIN_GROUP = new web3_js_1.PublicKey('78b8f4cGCwmZ9ysPFMWLaLTkkaYnUjwMJYStWe5RTSSX');
exports.USDC_MINT = new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v');
exports.MAX_RECENT_PRIORITY_FEE_ACCOUNTS = 128;
exports.JUPITER = {
    V3: new web3_js_1.PublicKey('JUP3c2Uh3WA4Ng34tw6kPd2G4C5BB21Xo36Je1s32Ph'),
    V4: new web3_js_1.PublicKey('JUP4Fb2cqiRUcaTHdrPC8h2gNsA2ETXiPDD33WcGuJB'),
    V6: new web3_js_1.PublicKey('JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4'),
};
exports.DEFAULT_RECEIVER_PROGRAM_ID = new web3_js_1.PublicKey('rec5EKMGg6MxZYaMdyBfgwp4d5rB9T1VQH5pJv5LtFJ');
// see maintain-alts.ts
exports.SBOD_ORACLE_LUTS = [
    '3DdohDpFiXjLNqCzFyU8CP3rtGwkn9VbkxFYGrNiqcyR',
];
