// @kamino-finance/limo-sdk/dist/rpc_client/types/UpdateGlobalConfigValue.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pubkey = exports.U64 = exports.U16 = exports.Bool = void 0;
exports.fromDecoded = fromDecoded;
exports.fromJSON = fromJSON;
exports.layout = layout;
const web3_js_1 = require("@solana/web3.js"); // eslint-disable-line @typescript-eslint/no-unused-vars
const bn_js_1 = __importDefault(require("bn.js")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class Bool {
    static discriminator = 0;
    static kind = "Bool";
    discriminator = 0;
    kind = "Bool";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "Bool",
            value: [this.value[0]],
        };
    }
    toEncodable() {
        return {
            Bool: {
                _0: this.value[0],
            },
        };
    }
}
exports.Bool = Bool;
class U16 {
    static discriminator = 1;
    static kind = "U16";
    discriminator = 1;
    kind = "U16";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "U16",
            value: [this.value[0]],
        };
    }
    toEncodable() {
        return {
            U16: {
                _0: this.value[0],
            },
        };
    }
}
exports.U16 = U16;
class U64 {
    static discriminator = 2;
    static kind = "U64";
    discriminator = 2;
    kind = "U64";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "U64",
            value: [this.value[0].toString()],
        };
    }
    toEncodable() {
        return {
            U64: {
                _0: this.value[0],
            },
        };
    }
}
exports.U64 = U64;
class Pubkey {
    static discriminator = 3;
    static kind = "Pubkey";
    discriminator = 3;
    kind = "Pubkey";
    value;
    constructor(value) {
        this.value = [value[0]];
    }
    toJSON() {
        return {
            kind: "Pubkey",
            value: [this.value[0].toString()],
        };
    }
    toEncodable() {
        return {
            Pubkey: {
                _0: this.value[0],
            },
        };
    }
}
exports.Pubkey = Pubkey;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("Bool" in obj) {
        const val = obj["Bool"];
        return new Bool([val["_0"]]);
    }
    if ("U16" in obj) {
        const val = obj["U16"];
        return new U16([val["_0"]]);
    }
    if ("U64" in obj) {
        const val = obj["U64"];
        return new U64([val["_0"]]);
    }
    if ("Pubkey" in obj) {
        const val = obj["Pubkey"];
        return new Pubkey([val["_0"]]);
    }
    throw new Error("Invalid enum object");
}
function fromJSON(obj) {
    switch (obj.kind) {
        case "Bool": {
            return new Bool([obj.value[0]]);
        }
        case "U16": {
            return new U16([obj.value[0]]);
        }
        case "U64": {
            return new U64([new bn_js_1.default(obj.value[0])]);
        }
        case "Pubkey": {
            return new Pubkey([new web3_js_1.PublicKey(obj.value[0])]);
        }
    }
}
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([borsh.bool("_0")], "Bool"),
        borsh.struct([borsh.u16("_0")], "U16"),
        borsh.struct([borsh.u64("_0")], "U64"),
        borsh.struct([borsh.publicKey("_0")], "Pubkey"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
